[DPE-4847] Adding documentation on Azure Storage backends (#35)
canonical/spark-k8s-bundle: releases/3.4/yaml/docs/how-to/h-setup-k8s.md

> ## How to setup a K8s cluster for Spark

Reviewer: Why do we start with h2? A page title usually is in h1 heading.

Author: I believe it is historical stuff, for which documentation looked "nicer" in charmhub using the h2. 

As discussed, we could address this in one ticket where we fix this consistently across all pages.

The Charmed Spark solution requires an environment with:

* A Kubernetes cluster for running the services and workloads
* An Object storage layer to store persistent data

> In the following, we provide details on the technologies currently supported and some further information on how these layers can be setup.

Reviewer: ```suggestion
In the following guide, we provide details on the technologies currently supported and instructions on how these layers can be set up.
```

> ### Kubernetes

Reviewer: Do we need this heading at all?

Author: I believe so, but there is an error down in the file [here](https://github.com/canonical/spark-k8s-bundle/blob/318da8eb7b581fa2d716331e8b92b1e5754a034d/releases/3.4/yaml/docs/how-to/h-setup-k8s.md?plain=1#L150), where we should only have 3 `#`. The idea is to have a TOC like:

* Kubernetes
   * MicroK8s
   * AWS EKS
* Object Storage
   * S3-compatible
      * MinIO
      * AWS S3
   * Azure storage

The Charmed Spark solution runs on top of several K8s distribution. We recommend you to use 1.29+, but no less than version 1.28. 
Earlier versions may still be working, although we do not explicitly test them. 

There are multiple ways that a K8s cluster can be deployed. We provide full compatibility and support for:

* MicroK8s
* AWS EKS

> * Azure AKS (**Coming Soon**)

Reviewer 2: Nice :)

The EKS cluster is now ready to be used. 

#### Object storage

Object storage persistence integration with Charmed Spark is critical for: 

* reading and writing application data to be used in Spark Jobs
* storing Spark Jobs logs to be then exposed via Charmed Spark History Server
* enable Hive-compatible JDBC/ODBC endpoints provided by Apache Kyuubi to provide datalake capabilities on top of HDFS/Hadoop/object storages

Charmed Spark provides out-of-box integration with the following object storage backends:

* S3-compatible object storages, such as:
  * MinIO
  * AWS S3 bucket
* Azure Storage 
  * Azure Blob Storage 
  * Azure DataLake v2 Storage

In the following, we provide guidance on how to set up different object storages and 
how to configure them to make sure that it seamlessly integrates with Charmed Spark. 

In fact, to store Spark logs on a dedicated directories, you need to create the appropriate folder (that is named `spark-events`) in the storage backend. 
This can be done both on S3 and on Azure DataLake Gen2 Storage. Although there are multiple ways to do this, 
in the following we recommend you to use snap clients. 

> Other possibilities could also be to use [Python libraries](https://github.com/canonical/spark-k8s-bundle/blob/94ac51d519cece9dc6810c7aaa0144a28cd7989b/python/spark_test/core/s3.py).

Reviewer: ```suggestion
Alternatively, use [Python libraries](https://github.com/canonical/spark-k8s-bundle/blob/94ac51d519cece9dc6810c7aaa0144a28cd7989b/python/spark_test/core/s3.py).
```
Is the link correct? It looks like a very specific commit.

#### S3-compatible object storages

> In order to connect Charmed Spark with an S3-compatible object storage, 

Reviewer: ```suggestion
To connect Charmed Spark with an S3-compatible object storage, 
```

##### Setting Up the object Storage

Leveraging on standard S3 API, you can use the `aws` snap client to perform operations with the
S3 service, like creating buckets, upload new content, inspecting the structure and removing data.

To install the AWS CLI client, use 

```shell
sudo snap install aws-cli --classic
```

The client can then be configured using the parameters above with 

```shell
aws configure set aws_access_key_id <S3_ACCESS_KEY>
aws configure set aws_secret_access_key <S3_SECRET_KEY>
aws configure set endpoint_url <S3_ENDPOINT>
aws configure set default.region <S3_REGION>
```

Test that the `aws-cli` client is properly working with 

```
aws s3 ls
```

To create a folder on an existing bucket, just place an empty path object `spark-events`. 

```shell
aws s3api put-object --bucket <S3_BUCKET> --key spark-events
```

The S3-object storage should now be ready to be used by Spark Jobs to store their logs. 

#### Azure Storage

Charmed Spark provides out-of-the-box support also for the following Azure storage backends: 

* Azure Blob Storage (both WASB and WASBS)
* Azure DataLake Gen2 Storage (ABFS and ABFSS)

> > :warning: Note that Azure DataLake Gen1 Storage is currently not supported, and it has been deprecated by 
> Azure.

Reviewer: ```suggestion
> :warning: Note that Azure DataLake Gen1 Storage is currently not supported, and it has been deprecated by Azure.
```
The whole content of the admonition block should be added to the quotation syntax (every line starts with `> `.
Secondly, I'm not sure how the `:warning:` piece is rendered in our docs. Is it an emoji?

Author: The `:warning:` renders as the triangle, yellow sign with the exclamation mark in the center. We use this elsewhere in the spark docs (e.g. how to deploy), and it is also used in other docs.

In order to connect Charmed Spark with the Azure storage backends (WASB, WASBS, ABFS and ABFSS) 
the following configurations need to be specified:

* *storage_account*
* *storage_key*
* *container*

##### Setting Up the object Storage

You can use the `azcli` snap client to perform operations with the Azure storage services, 
like creating buckets, upload new content, inspecting the structure and removing data.

To install the `azcli` client, use 

```shell
sudo snap install azcli
```

The client can then be configured using the following environment variables

```shell
export AZURE_STORAGE_ACCOUNT=...
export AZURE_STORAGE_KEY=...
```

These credentials can be retrieved from the [Azure portal](https://portal.azure.com/#view/HubsExtension/BrowseResource/resourceType/Microsoft.Storage%2FStorageAccounts), after creating a storage account. 
When creating the storage account, make sure that you enable "Hierarchical namespace" if you 
want to use Azure DataLake Gen2 Storage.

> Test that the `aws-cli` client is properly working with 

Reviewer: Is it `aws-cli` or `azcli`?

Author: yes, that was a typo. Great catch!
